File: python/a2021/18.py
from problem import get_problem_lines

def tosn(s):
    return [int(c) if c.isdigit() else c for c in s]

def psn(sn):
    for c in sn:
        print(c,end='')
    print()

def explode(sn):
    depth = 0
    k = None
    for i in range(len(sn)):
        if sn[i] == '[':
            depth += 1
        if sn[i] == ']':
            depth -= 1
        if depth > 4 and isinstance(sn[i],int) and sn[i+1] == ',' and isinstance(sn[i+2],int) :
            k = i
            break
    if k == None:
        return (False, sn)
    for j in range(k-1,-1,-1):
        if isinstance(sn[j], int):
            sn[j] += sn[k]
            break
    for j in range(k+3,len(sn)):
        if isinstance(sn[j], int):
            sn[j] += sn[k+2]
            break
    sn = sn[:k-1] + [0] + sn[k+4:]
    return (True, sn)

def split(sn):
    k = None
    for (i, c) in enumerate(sn):
        if isinstance(c, int) and c >= 10:
            k = i
            break
    if k == None:
        return (False, sn)
    l, r = sn[k] // 2, (sn[k] - (sn[k] // 2 ))
    sn = sn[:k] + ['[',l ,',', r, ']'] + sn[k+1:]
    return (True, sn)

def add(sn1, sn2):
    return ['['] + sn1 + [','] + sn2 + [']']

def simplify(sn):
    while True:
        c, sn = explode(sn)
        if c:
            continue
        c, sn = split(sn)
        if not c:
            return sn

def magnitude(sn):
    k = 1
    ans = 0
    for c in sn:
        if c == '[':
            k = k * 3
        if c == ',':
            k = k // 3
            k = k * 2
        if c == ']':
            k = k // 2
        if isinstance(c, int):
            ans += k*c
    return ans

lines = [simplify(tosn(x)) for x in get_problem_lines()]


sn = lines[0]
for line in lines[1:]:
    sn = simplify(add(sn, line))
print(f'part 1: {magnitude(sn)}')

ans = 0
for i in range(len(lines)):
    for j in range(len(lines)):
        if i == j:
            continue
        ans = max(ans, magnitude(simplify(add(lines[i],lines[j]))))

print(f'part 2: {ans}')

File: rust/a2021/src/bin/18.rs
#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum SNode {
    Num(u8),
    LBra,
    Comma,
    RBra,
}

type SVec = Vec<SNode>;

fn tosn(s: &str) -> SVec {
    s.chars()
        .map(|c| match c {
            '[' => SNode::LBra,
            ',' => SNode::Comma,
            ']' => SNode::RBra,
            c => SNode::Num(c.to_digit(10).unwrap() as u8),
        })
        .collect::<Vec<SNode>>()
}

/*
fn psn(s: &SVec) {
    for c in s {
        match c {
            SNode::LBra => print!("["),
            SNode::Comma => print!(","),
            SNode::RBra => print!("]"),
            SNode::Num(n) => print!("{n}"),
        }
    }
    println!("");
}
*/

fn explode(sn: &mut SVec) -> bool {
    let mut depth = 0;
    let mut k = None;
    for i in 0..sn.len() {
        if sn[i] == SNode::LBra {
            depth += 1;
        } else if sn[i] == SNode::RBra {
            depth -= 1;
        } else if depth > 4 {
            if let SNode::Num(l) = sn[i] {
                if let SNode::Num(r) = sn[i + 2] {
                    k = Some((i, l, r));
                    break;
                }
            }
        }
    }
    if let Some((i, l, r)) = k {
        for p in sn[..i].iter_mut().rev() {
            if let SNode::Num(v) = *p {
                *p = SNode::Num(v + l);
                break;
            }
        }
        for p in &mut sn[i + 3..] {
            if let SNode::Num(v) = *p {
                *p = SNode::Num(v + r);
                break;
            }
        }
        sn[i - 1] = SNode::Num(0);
        sn.drain(i..=i + 3);
        true
    } else {
        false
    }
}

fn split(sn: &mut SVec) -> bool {
    let mut k = None;
    for (i, p) in sn.iter().enumerate() {
        if let &SNode::Num(v) = p {
            if v >= 10 {
                k = Some((i, v));
                break;
            }
        }
    }
    if let Some((i, v)) = k {
        let l = v / 2;
        let r = v - l;
        sn.splice(
            i..=i,
            vec![
                SNode::LBra,
                SNode::Num(l),
                SNode::Comma,
                SNode::Num(r),
                SNode::RBra,
            ],
        );
        true
    } else {
        false
    }
}

fn add(sn1: &SVec, sn2: &SVec) -> SVec {
    let mut a: Vec<SNode> = vec![SNode::LBra];
    a.extend(sn1);
    a.push(SNode::Comma);
    a.extend(sn2);
    a.push(SNode::RBra);
    a
}

fn simplify(sn: &mut SVec) {
    while explode(sn) || split(sn) {}
}

fn magnitude(sn: &SVec) -> u32 {
    let mut k = 1;
    let mut ans = 0;
    for c in sn {
        match c {
            SNode::LBra => k *= 3,
            SNode::Comma => {
                k /= 3;
                k *= 2;
            }
            SNode::RBra => k /= 2,
            SNode::Num(c) => ans += (*c as u32) * k,
        }
    }
    ans
}

fn main() {
    let mut lines = vec![];
    for line in common::get_problem_lines(2021, 18) {
        lines.push(tosn(&line));
    }

    let mut sn = lines[0].clone();
    for sn_ in &lines[1..] {
        sn = add(&sn, sn_);
        simplify(&mut sn);
    }
    println!("part 1: {}", magnitude(&sn));
    explode(&mut sn);

    let mut ans = 0;
    for i in 0..lines.len() {
        for j in 0..lines.len() {
            if i == j {
                continue;
            }
            let mut sn = add(&lines[i], &lines[j]);
            simplify(&mut sn);
            ans = ans.max(magnitude(&sn));
        }
    }
    println!("part 2: {}", ans);
}
